include/LightGBM/category_encoding_provider.hpp

```diff
@@ -0,0 +1,771 @@
+/*!
+  * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+  * Licensed under the MIT License. See LICENSE file in the project root for license information.
+  */
+#ifndef LIGHTGBM_CATEGORY_ENCODING_PROVIDER_HPP_
+#define LIGHTGBM_CATEGORY_ENCODING_PROVIDER_HPP_
+
+#include <LightGBM/config.h>
+#include <LightGBM/network.h>
+#include <LightGBM/utils/json11.h>
+#include <LightGBM/utils/text_reader.h>
+#include <LightGBM/utils/threading.h>
+#include <LightGBM/parser_base.h>
+
+#include <algorithm>
+#include <memory>
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+namespace LightGBM {
+
+using json11::Json;
+
+// transform categorical features to encoded numerical values before the bin construction process
+class CategoryEncodingProvider {
+ public:
+  class CatConverter {
+   protected:
+    std::unordered_map<int, int> cat_fid_to_convert_fid_;
+
+   public:
+    virtual ~CatConverter() {}
+
+    virtual double CalcValue(const double sum_label, const double sum_count,
+      const double all_fold_sum_count, const double prior) const = 0;
+
+    virtual double CalcValue(const double sum_label, const double sum_count,
+      const double all_fold_sum_count) const = 0;
+
+    virtual std::string DumpToString() const = 0;
+
+    virtual json11::Json DumpToJSONObject() const = 0;
```

Comment:
To make sure we have common pattern for serialization could we use json as general pattern? json is a special string then we do not need to define our customize pattern?

Reply:
And I believe we should not do serialization & deserialization very often?